BvgFetcher.js

```diff
@@ -4,9 +4,7 @@ const bvgProfile = require("hafas-client/p/bvg");
 const pjson = require("./package.json");
 const bvgClient = createClient(
   bvgProfile,
-  "MagicMirror2 module MMM-PublicTransportBerlin v" +
-    pjson.version +
-    " (https://github.com/deg0nz/MMM-PublicTransportBerlin)"
+  "MMM-PublicTransportBerlin v" + pjson.version
```

Comment:
I suggest to add a installation-specific part to the user agent. This way, admins of the HAFAS API can see that the load they get from Magic Mirrors comes from *mutiple* people/installations, not just one person.

Reply:
Yeah, that's a good idea. But won't we come back to the original problem regarding the long user agent string? 

The MagicMirror creates [`this.identifier` for every module as unique ID](https://docs.magicmirror.builders/development/core-module-file.html#available-module-instance-properties), but as it seems,[ this is only a string containing the index in the module list](https://github.com/MichMich/MagicMirror/blob/26a76f80d63e5c9746b6db712d6901219f8b6b93/js/loader.js#L101).

Reply:
From my experience, a too long user-agent is not a problem, but I haven't specifically looked into it.

Reply:
The requests come from different IP addresses. Shouldn't this be obvious enough for the admins that there are different users?

But I am not against adding anything specific to the user agent string :slightly_smiling_face: This could perhaps also be a randomly generated session ID that is created each time the module/mirror is started.

Reply:
> The requests come from different IP addresses. Shouldn't this be obvious enough for the admins that there are different users?

I don't know. But they haven't made the impression of picking effective measures to me. 😄

> But I am not against adding anything specific to the user agent string 🙂 This could perhaps also be a randomly generated session ID that is created each time the module/mirror is started.

Just FYI: [`hafas-client` also randomises the user-agent with every request](https://github.com/public-transport/hafas-client/blob/2ec079adfc8a3d988190491b7e07dc03826b719e/lib/request.js#L88).

Reply:
> > But I am not against adding anything specific to the user agent string 🙂 This could perhaps also be a randomly generated session ID that is created each time the module/mirror is started.
> 
> Just FYI: [`hafas-client` also randomises the user-agent with every request](https://github.com/public-transport/hafas-client/blob/2ec079adfc8a3d988190491b7e07dc03826b719e/lib/request.js#L88).

Ah, maybe a long user-agent *does* make a difference, because a short string will likely be split by `randomizeUserAgent` into 2 reasonably short segments that can't easily be blocked; With increasing length, the likelihood of a long – and therefore unique – segment being among the 2 segments is getting higher.

Reply:
> Just FYI: [`hafas-client` also randomises the user-agent with every request](https://github.com/public-transport/hafas-client/blob/2ec079adfc8a3d988190491b7e07dc03826b719e/lib/request.js#L88).

This is interesting :slightly_smiling_face: 

I disabled the randomization process and found the following: As soon as the user agent contains the string `MMM-PublicTransportBerlin`, the query does not work. No matter how long the string is.

So if the hafas-client does not touch this part in its randomization process, the error message appears. This explains why it mostly works with the shorter string (because it is more likely that the string will be changed).

Apparently the queries are not wanted on the endpoint's side. Would it make sense to ask the BVG why they installed the filter? Are there other sources for the data besides the BVG hafas endpoint?

Reply:
Wow, that's definitely a finding.

I could imagine that they have some kind of rating limiter that seems to check for user agents. Fail2ban can be configured that way afaik. 

I could imagine that there are some instances running...

> Are there other sources for the data besides the BVG hafas endpoint?

We could switch back to the VBB-Endpoint, which I wanted to do for some time now. The `hafas-client` with BVG profile was only active because the VBB APIs were very unreliable some years (is it that long ?!) ago.

But I don't know the internals of the hafas APIs, maybe they will block the user agents globally for all profiles.

Reply:
Thanks for investigating this. I'll publish a modified randomisation logic today that doesn't have this problem.

Reply:
> > Are there other sources for the data besides the BVG hafas endpoint?
> 
> We could switch back to the VBB-Endpoint, which I wanted to do for some time now. The `hafas-client` with BVG profile was only active because the VBB APIs were very unreliable some years (is it that long ?!) ago.

The VBB endpoint also has a rate limiting in place. By now, AFAICT, most if not all HAFAS endpoints do.

Reply:
I just published `hafas-client@5.22.2`, which contains a new randomisation logic that inserts a random hex string every 5-10 characters.

Reply:
Noice. Thanks for the quick actions! I will merge this later today with an updated `hafas-client`. Let's hope this works 🤞

Reply:
I did the update. Looks good :smiley: 

Thanks @derhuerst :+1: 

But it bothers me that we have to randomize the user-agent at all.